File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/sync/ClusterSyncFlowConfig.java
package com.sequenceiq.cloudbreak.core.flow2.cluster.sync;

import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.CLUSTER_SYNC_EVENT;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FAILURE_EVENT;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FAIL_HANDLED_EVENT;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FINALIZED_EVENT;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.CLUSTER_SYNC_FAILED_STATE;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.CLUSTER_SYNC_STATE;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.FINAL_STATE;
import static com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.INIT_STATE;

import java.util.List;

import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.core.flow2.config.AbstractFlowConfiguration;
import com.sequenceiq.cloudbreak.core.flow2.config.AbstractFlowConfiguration.Transition.Builder;
import com.sequenceiq.cloudbreak.core.flow2.config.RetryableFlowConfiguration;

@Component
public class ClusterSyncFlowConfig extends AbstractFlowConfiguration<ClusterSyncState, ClusterSyncEvent>
    implements RetryableFlowConfiguration<ClusterSyncEvent> {

    private static final List<Transition<ClusterSyncState, ClusterSyncEvent>> TRANSITIONS =
            new Builder<ClusterSyncState, ClusterSyncEvent>()
                    .from(INIT_STATE).to(CLUSTER_SYNC_STATE).event(CLUSTER_SYNC_EVENT).noFailureEvent()
                    .from(CLUSTER_SYNC_STATE).to(ClusterSyncState.CLUSTER_SYNC_FINISHED_STATE).event(ClusterSyncEvent.CLUSTER_SYNC_FINISHED_EVENT)
                    .failureEvent(ClusterSyncEvent.CLUSTER_SYNC_FINISHED_FAILURE_EVENT)
                    .from(ClusterSyncState.CLUSTER_SYNC_FINISHED_STATE).to(FINAL_STATE).event(FINALIZED_EVENT).failureEvent(FAILURE_EVENT)
                    .build();

    private static final FlowEdgeConfig<ClusterSyncState, ClusterSyncEvent> EDGE_CONFIG = new FlowEdgeConfig<>(INIT_STATE, FINAL_STATE,
            CLUSTER_SYNC_FAILED_STATE, FAIL_HANDLED_EVENT);

    private static final String FLOW_DISPLAY_NAME = "Cluster Sync";

    public ClusterSyncFlowConfig() {
        super(ClusterSyncState.class, ClusterSyncEvent.class);
    }

    @Override
    protected List<Transition<ClusterSyncState, ClusterSyncEvent>> getTransitions() {
        return TRANSITIONS;
    }

    @Override
    protected FlowEdgeConfig<ClusterSyncState, ClusterSyncEvent> getEdgeConfig() {
        return EDGE_CONFIG;
    }

    @Override
    public String getDisplayName() {
        return FLOW_DISPLAY_NAME;
    }

    @Override
    public ClusterSyncEvent[] getEvents() {
        return ClusterSyncEvent.values();
    }

    @Override
    public ClusterSyncEvent[] getInitEvents() {
        return new ClusterSyncEvent[]{
                CLUSTER_SYNC_EVENT
        };
    }

    @Override
    public ClusterSyncEvent getFailHandledEvent() {
        return FAIL_HANDLED_EVENT;
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/sync/ClusterSyncFlowConfig.kt
package com.sequenceiq.cloudbreak.core.flow2.cluster.sync

import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.CLUSTER_SYNC_EVENT
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FAILURE_EVENT
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FAIL_HANDLED_EVENT
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncEvent.FINALIZED_EVENT
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.CLUSTER_SYNC_FAILED_STATE
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.CLUSTER_SYNC_STATE
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.FINAL_STATE
import com.sequenceiq.cloudbreak.core.flow2.cluster.sync.ClusterSyncState.INIT_STATE

import org.springframework.stereotype.Component

import com.sequenceiq.cloudbreak.core.flow2.config.AbstractFlowConfiguration

@Component
class ClusterSyncFlowConfig : AbstractFlowConfiguration<ClusterSyncState, ClusterSyncEvent>(ClusterSyncState::class.java, ClusterSyncEvent::class.java) {

    protected override val transitions: List<AbstractFlowConfiguration.Transition<ClusterSyncState, ClusterSyncEvent>>
        get() = TRANSITIONS

    protected override val edgeConfig: AbstractFlowConfiguration.FlowEdgeConfig<ClusterSyncState, ClusterSyncEvent>
        get() = EDGE_CONFIG

    override val events: Array<ClusterSyncEvent>
        get() = ClusterSyncEvent.values()

    override val initEvents: Array<ClusterSyncEvent>
        get() = arrayOf(CLUSTER_SYNC_EVENT)

    companion object {
        private val TRANSITIONS = Transition.Builder<ClusterSyncState, ClusterSyncEvent>().from(INIT_STATE).to(CLUSTER_SYNC_STATE).event(CLUSTER_SYNC_EVENT).noFailureEvent().from(CLUSTER_SYNC_STATE).to(ClusterSyncState.CLUSTER_SYNC_FINISHED_STATE).event(ClusterSyncEvent.CLUSTER_SYNC_FINISHED_EVENT).failureEvent(ClusterSyncEvent.CLUSTER_SYNC_FINISHED_FAILURE_EVENT).from(ClusterSyncState.CLUSTER_SYNC_FINISHED_STATE).to(FINAL_STATE).event(FINALIZED_EVENT).failureEvent(FAILURE_EVENT).build()

        private val EDGE_CONFIG = AbstractFlowConfiguration.FlowEdgeConfig(INIT_STATE, FINAL_STATE,
                CLUSTER_SYNC_FAILED_STATE, FAIL_HANDLED_EVENT)
    }
}
